crates/bevy_picking/src/mesh_picking/ray_cast/intersections.rs

```diff
@@ -0,0 +1,405 @@
+    pub triangle_index: Option<usize>,
+}
+
+/// A hit result from a ray cast on a triangle.
+#[derive(Default, Debug)]
+pub struct RayTriangleHit {
+    pub distance: f32,
+    pub barycentric_coords: (f32, f32),
+}
+
+/// Casts a ray on a mesh, and returns the intersection.
+pub(super) fn ray_intersection_over_mesh(
+    mesh: &Mesh,
+    mesh_transform: &Mat4,
+    ray: Ray3d,
+    backface_culling: Backfaces,
+) -> Option<RayMeshHit> {
+    if mesh.primitive_topology() != PrimitiveTopology::TriangleList {
+        error!(
+            "Invalid intersection check: `TriangleList` is the only supported `PrimitiveTopology`"
+        );
+        return None;
+    }
+
+    // Get the vertex positions and normals from the mesh.
+    let vertex_positions: &Vec<[f32; 3]> = match mesh.attribute(Mesh::ATTRIBUTE_POSITION) {
+        None => panic!("Mesh does not contain vertex positions"),
+        Some(vertex_values) => match &vertex_values {
+            VertexAttributeValues::Float32x3(positions) => positions,
+            _ => panic!("Unexpected types in {:?}", Mesh::ATTRIBUTE_POSITION),
+        },
+    };
+    let vertex_normals: Option<&[[f32; 3]]> =
+        if let Some(normal_values) = mesh.attribute(Mesh::ATTRIBUTE_NORMAL) {
+            match &normal_values {
+                VertexAttributeValues::Float32x3(normals) => Some(normals),
+                _ => None,
+            }
+        } else {
+            None
+        };
+
+    if let Some(indices) = &mesh.indices() {
+        match indices {
+            Indices::U16(vertex_indices) => ray_mesh_intersection(
+                ray,
+                mesh_transform,
+                vertex_positions,
+                vertex_normals,
+                Some(vertex_indices),
+                backface_culling,
+            ),
+            Indices::U32(vertex_indices) => ray_mesh_intersection(
+                ray,
+                mesh_transform,
+                vertex_positions,
+                vertex_normals,
+                Some(vertex_indices),
+                backface_culling,
+            ),
+        }
+    } else {
+        ray_mesh_intersection(
+            ray,
+            mesh_transform,
+            vertex_positions,
+            vertex_normals,
+            None::<&Vec<u32>>,
+            backface_culling,
+        )
+    }
+}
+
+/// A trait for converting a value into a [`usize`].
+pub trait IntoUsize: Copy {
+    /// Converts the value into a [`usize`].
+    fn into_usize(self) -> usize;
+}
+
+impl IntoUsize for u16 {
+    fn into_usize(self) -> usize {
+        self as usize
+    }
+}
+
+impl IntoUsize for u32 {
+    fn into_usize(self) -> usize {
+        self as usize
+    }
+}
+
+/// Checks if a ray intersects a mesh, and returns the nearest intersection if one exists.
+pub fn ray_mesh_intersection(
+    ray: Ray3d,
+    mesh_transform: &Mat4,
+    vertex_positions: &[[f32; 3]],
+    vertex_normals: Option<&[[f32; 3]]>,
+    indices: Option<&Vec<impl IntoUsize>>,
```

Comment:
```suggestion
/// Checks if a ray intersects a mesh, and returns the nearest intersection if one exists.
pub fn ray_mesh_intersection(
    ray: Ray3d,
    mesh_transform: &Mat4,
    vertex_positions: &[[f32; 3]],
    vertex_normals: Option<&[[f32; 3]]>,
    indices: Option<&Vec<impl Into<usize> + Copy>>,
```

The trait `IntoUsize` seems redundant, since `usize` implements `From<u32>` and `From<u16>` anyway.

Reply:
It does implement `From<u16>`, but not `From<u32>`, because that isn't guaranteed to be a lossless conversion since `u32` doesn't fit inside a `usize` on all platforms (like some microcontrollers with a native size of `u16`). I believe `as` simply truncates the number so that it fits in those cases.

It *does* implement `TryFrom<u32>` though, which afaik returns an error when the number doesn't fit. I changed it to use this for now, didn't see any meaningful performance change in benchmarks.